File: dfs/572.subtree-of-another-tree.go
package leetcode

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSame(t1 *TreeNode, t2 *TreeNode) bool {
	if t1 == nil && t2 == nil {
		return true
	}
	if t1 == nil || t2 == nil {
		return false
	}
	if t1.Val == t2.Val && isSame(t1.Left, t2.Left) && isSame(t1.Right, t2.Right) {
		return true
	}
	return false
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
  if root == nil {
    return false
  }
  if isSame(root, subRoot){
    return true
  }
  if isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot) {
    return true
  }
  // don't manuall check child node. use recursive
  // if isSame(root.Left, subRoot) {
  //   return true
  // }
  // if isSame(root.Right, subRoot) {
  //   return true
  // }
  return false
}

// @leet end

File: dfs/572.subtree-of-another-tree.ts
import { TreeNode } from "./type";

// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function check(t1: TreeNode | null, t2: TreeNode | null): boolean {
  if (t1 === null && t2 === null) return true;
  if (t1 === null || t2 === null) return false;
  if ( t1.val === t2.val && check(t1.left, t2.left) && check(t1.right, t2.right)){
    return true
  }
  return false
}
function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
  if ( root === null) return false
  if (check(root, subRoot)) return true
  return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
}
// @leet end

